components/feature/media/src/main/java/mozilla/components/feature/media/middleware/InProgressMediaMiddleware.kt

```diff
@@ -0,0 +1,36 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.media.middleware
+
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.InProgressMediaAction
+import mozilla.components.browser.state.action.MediaSessionAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.concept.engine.mediasession.MediaSession
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+
+/**
+ * [Middleware] that updates [BrowserState.inProgressMediaTabId] everytime the user start playing media
+ * or navigates to another page in the last tab that was playing media.
+ */
+class InProgressMediaMiddleware : Middleware<BrowserState, BrowserAction> {
+    @Suppress("ComplexCondition")
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        next(action)
+
+        if (action is MediaSessionAction.UpdateMediaPlaybackStateAction &&
+            action.playbackState == MediaSession.PlaybackState.PLAYING) {
+
+            context.dispatch(InProgressMediaAction.PlayStarted(action.tabId))
+        } else if (action is MediaSessionAction.DeactivatedMediaSessionAction) {
```

Comment:
@pocmo and I were wondering if this is good enough, which is why we suggested the combination with `url changed`. If the user stops/pauses the media will this trigger? Because then the feature won't work, right? We set `lastMediaAccess` to 0 and then wouldn't display it?

Reply:
I guess in other words, I am not sure when `DeactivatedMediaSessionAction` fires?

Reply:
Using `soundcloud` I see indeed that if after playing media I tap on other links in the page `UpdateUrlAction` fires but media continues to play. So this cannot be trusted.
`DeactivatedMediaSessionAction` seems to fire exactly in the situation we wanted: user navigates to another page and GeckoView closes the current MediaSession.
Will ask on `Matrix - GeckoView` for more details about `DeactivatedMediaSessionAction` and if we can use it as an indicative that the user navigated to another page.

Reply:
OK, I doubt we can cover all cases, but probably don't need to. In general, if the user stops/pauses, switches tabs or closes the app we would like to be able to show the media on home. If they navigated away, as you wrote, we don't need to, but it's not clear how reliably we can detect that e.g. in soundcloud where you can navigate and the media keeps playing on the page. I guess a workaround could be navigated to a different domain?

Reply:
`alwu` [responded on Matrix that](https://matrix.to/#/!LfXZSWEroPFPMQcYmw:mozilla.org/$gR7lOUgErQ6NI7D76Sb-f9786ISo1GOQFcVnwuKBSnA?via=mozilla.org&via=matrix.org&via=matrix.start9labs.com):
> When users navigate to other url which is still the part of the Soundcloud, as it's still same document and the active media session still exists. So we would keep the GeckoView's MediaSession active. When navigating to the Wikipedia, that is a totally different page, and the MediaSession (via Web API) created by SoundCloud will be destroyed which will also cause GeckoView's MediaSession triggering deactivation.

Reply:
OK, sounds good! Let's start with that and iterate as needed.